### ISSUE TITLE
Too many ways / commands to add add files to the store manually

### ISSUE THREAD

**Issue by doronbehar** [NixOS/nix#6210]
This issue is a comment upon the design of some `nix` commands / lack of proper documentation to them. The commands are:

1. `nix store add-file`
3. `nix store add-path`
4. `nix-store --add`
5. `nix-store --add-fixed`
1. `nix-prefetch --requireFile`

I think I understood from the docs that (2) and (3) are meant for files and directories respectively, but why doesn't nix differentiate between a directory and a file by it's own? The same comment is relevant for `nix-store --add` vs `nix-store --add-fixed`. The naming of these options is not clear to me - Why would someone not use `--recursive` with `nix-store --add-fixed` for a path which is a directory? Can't it be used automatically when the path argument is a directory? Is using `--add` means the contents aren't added as fixed output?

In comparison between the `nix-store` commands and the `nix store add-` commands: How does `nix` compute the hash for the path arguments I give to (2) and (3)? Perhaps it'd have been best to print the hash as well? Or add a `--json` option that would print the hash? Why does `nix-store --add-fixed` doesn't accept a `--name` argument like `nix store add-path`?

Lastly, is it possible to use `nix-prefetch --requireFile` instead of any of these commands?

**Comment by edolstra** [comment]
> Not really sure what the flat hashing scheme is useful for

It's the hashing scheme used by fixed-output derivations like `fetchurl`. So `nix store add-file` will produce the same store path as `fetchurl`.

**Comment by doronbehar** [comment]
So is there any reason someone would want to use the recursive hashing scheme for single files? for tarballs downloaded with `fetchurl`?

**Comment by edolstra** [comment]
Generally no. The main reason for using the recursive scheme for single files is that it supports the executable bit.

**Comment by doronbehar** [comment]
So if someone wishes to add an executable to their store, they can use the flat hashing scheme and then `chmod +x` that executable right?

**Comment by edolstra** [comment]
No, because the store is read-only, so you can't (and shouldn't) change the `x` bit on files.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
